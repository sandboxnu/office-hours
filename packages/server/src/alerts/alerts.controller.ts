import {
  CreateAlertParams,
  CreateAlertResponse,
  ERROR_MESSAGES,
  GetAlertsResponse,
  Role,
} from '@koh/common';
import {
  BadRequestException,
  Body,
  Controller,
  Get,
  Param,
  Patch,
  Post,
  UseGuards,
} from '@nestjs/common';
import { JwtAuthGuard } from 'guards/jwt-auth.guard';
import { User } from 'decorators/user.decorator';
import { UserModel } from 'profile/user.entity';
import { Roles } from '../decorators/roles.decorator';
import { AlertModel } from './alerts.entity';
import { AlertsService } from './alerts.service';

@Controller('alerts')
@UseGuards(JwtAuthGuard)
export class AlertsController {
  constructor(private alertsService: AlertsService) {}

  @Get(':courseId')
  async getAlerts(
    @Param('courseId') courseId: number,
    @User() user: UserModel,
  ): Promise<GetAlertsResponse> {
    const alerts = await AlertModel.find({
      where: {
        courseId,
        user,
        resolved: null,
      },
    });
    return { alerts: await this.alertsService.removeStaleAlerts(alerts) };
  }

  @Post()
  @Roles(Role.TA, Role.PROFESSOR)
  async createAlert(
    @Body() body: CreateAlertParams,
  ): Promise<CreateAlertResponse> {
    const { alertType, courseId, payload, targetUserId } = body;

    if (!this.alertsService.assertPayloadType(alertType, payload)) {
      throw new BadRequestException(
        ERROR_MESSAGES.alertController.incorrectPayload,
      );
    }

    const anotherAlert = await AlertModel.findOne({
      where: {
        alertType,
        userId: targetUserId,
        resolved: null,
      },
    });

    // If the same user already has an alert for this then don't create a new one
    if (anotherAlert) {
      throw new BadRequestException(
        ERROR_MESSAGES.alertController.duplicateAlert,
      );
    }

    const alert = await AlertModel.create({
      alertType,
      sent: new Date(),
      userId: targetUserId,
      courseId,
      payload,
    }).save();

    return alert;
  }

  @Patch(':alertId')
  @Roles(Role.STUDENT, Role.TA, Role.PROFESSOR)
  async closeAlert(@Param('alertId') alertId: number): Promise<void> {
    const alert = await AlertModel.findOne({
      where: {
        id: alertId,
      },
    });

    if (!alert) {
      throw new BadRequestException(
        ERROR_MESSAGES.alertController.notActiveAlert,
      );
    }

    alert.resolved = new Date();
    await alert.save();
  }
}
