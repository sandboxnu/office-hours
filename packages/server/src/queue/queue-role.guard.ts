import { Injectable, UnauthorizedException } from '@nestjs/common';
import { UserModel } from '../profile/user.entity';
import { RolesGuard } from '../guards/role.guard';
import { QueueModel } from './queue.entity';

@Injectable()
export class QueueRolesGuard extends RolesGuard {
  // eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
  async setupData(
    request: any,
  ): Promise<{ courseId: number; user: UserModel }> {
    const queue = await QueueModel.findOne(request.params.queueId);
    const courseId = queue?.courseId;
    const user = await UserModel.findOne(request.user.userId, {
      relations: ['courses'],
    });

    return { courseId, user };
  }
}
