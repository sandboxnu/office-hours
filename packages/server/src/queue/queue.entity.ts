import { Exclude } from 'class-transformer';
import { QuestionGroupModel } from '../question/question-group.entity';
import {
  BaseEntity,
  Column,
  Entity,
  JoinColumn,
  JoinTable,
  ManyToMany,
  ManyToOne,
  OneToMany,
  PrimaryGeneratedColumn,
} from 'typeorm';
import { CourseModel } from '../course/course.entity';
import { UserModel } from '../profile/user.entity';
import { QuestionModel } from '../question/question.entity';

import { HttpException, HttpStatus } from '@nestjs/common';
import { ERROR_MESSAGES } from '@koh/common';

@Entity('queue_model')
export class QueueModel extends BaseEntity {
  @PrimaryGeneratedColumn()
  id: number;

  @ManyToOne((type) => CourseModel, (course) => course.queues)
  @JoinColumn({ name: 'courseId' })
  course: CourseModel;

  @Column({ nullable: true })
  @Exclude()
  courseId: number;

  @Column('text')
  room: string;

  @OneToMany((type) => QuestionModel, (qm) => qm.queue)
  @Exclude()
  questions: QuestionModel[];

  @OneToMany((type) => QuestionGroupModel, (qg) => qg.queue)
  @Exclude()
  groups: QuestionGroupModel[];

  @Column('text', { nullable: true })
  notes: string;

  @ManyToMany((type) => UserModel, (user) => user.queues)
  @JoinTable()
  staffList: UserModel[];

  @Column({ default: false })
  allowQuestions: boolean;

  @Column({ default: false })
  isProfessorQueue: boolean;

  @Column({ default: false })
  isDisabled: boolean;

  startTime: Date;
  endTime: Date;

  isOpen: boolean;

  async checkIsOpen(): Promise<boolean> {
    if (!this.staffList) {
      console.error(ERROR_MESSAGES.queueController.missingStaffList, this.id);
      throw new HttpException(
        ERROR_MESSAGES.queueController.missingStaffList,
        HttpStatus.INTERNAL_SERVER_ERROR,
      );
    }
    this.isOpen = this.staffList.length > 0 && !this.isDisabled;
    return this.isOpen;
  }

  queueSize: number;

  async addQueueSize(): Promise<void> {
    this.queueSize = await QuestionModel.waitingInQueue(this.id).getCount();
  }

  // TODO: eventually figure out how staff get sent to FE as well
}
