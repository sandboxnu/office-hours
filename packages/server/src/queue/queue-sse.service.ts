import { Injectable } from '@nestjs/common';
import { SSEService } from 'sse/sse.service';
import { QueueService } from './queue.service';
import { Response } from 'express';
import { throttle } from 'lodash';

const idToRoom = (queueId: number) => `q-${queueId}`;
/**
 * Handle sending queue sse events
 */
@Injectable()
export class QueueSSEService {
  constructor(
    private queueService: QueueService,
    private sseService: SSEService,
  ) {}

  subscribeClient(queueId: number, res: Response): void {
    this.sseService.subscribeClient(idToRoom(queueId), res);
  }

  // Send event with new questions, but no more than once a second
  updateQuestions = throttle(
    async (queueId: number) => {
      const questions = await this.queueService
        .getQuestions(queueId)
        .catch((e) => null);
      if (questions) {
        this.sseService.sendEvent(idToRoom(queueId), questions);
      }
    },
    1000,
    {
      leading: false,
      trailing: true,
    },
  );
}
