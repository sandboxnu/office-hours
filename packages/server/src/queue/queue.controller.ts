import {
  ERROR_MESSAGES,
  GetQueueResponse,
  ListQuestionsResponse,
  Role,
  UpdateQueueParams,
} from '@koh/common';
import {
  Body,
  ClassSerializerInterceptor,
  Controller,
  Delete,
  Get,
  HttpException,
  HttpStatus,
  NotFoundException,
  Param,
  Patch,
  Post,
  Res,
  UseGuards,
  UseInterceptors,
} from '@nestjs/common';
import { Response } from 'express';
import { UserId } from 'decorators/user.decorator';
import { Connection } from 'typeorm';
import { JwtAuthGuard } from '../guards/jwt-auth.guard';
import { Roles } from '../decorators/roles.decorator';
import { QueueCleanService } from './queue-clean/queue-clean.service';
import { QueueRole } from '../decorators/queue-role.decorator';
import { QueueRolesGuard } from '../guards/queue-role.guard';
import { QueueSSEService } from './queue-sse.service';
import { QueueModel } from './queue.entity';
import { QueueService } from './queue.service';

@Controller('queues')
@UseGuards(JwtAuthGuard, QueueRolesGuard)
@UseInterceptors(ClassSerializerInterceptor)
export class QueueController {
  constructor(
    private connection: Connection,
    private queueSSEService: QueueSSEService,
    private queueCleanService: QueueCleanService,
    private queueService: QueueService,
  ) {}

  @Get(':queueId')
  @Roles(Role.TA, Role.PROFESSOR, Role.STUDENT)
  async getQueue(@Param('queueId') queueId: number): Promise<GetQueueResponse> {
    try {
      return this.queueService.getQueue(queueId);
    } catch (err) {
      console.error(err);
      throw new HttpException(
        ERROR_MESSAGES.queueController.getQueue,
        HttpStatus.NOT_FOUND,
      );
    }
  }

  @Get(':queueId/questions')
  @Roles(Role.TA, Role.PROFESSOR, Role.STUDENT)
  async getQuestions(
    @Param('queueId') queueId: number,
    @QueueRole() role: Role,
    @UserId() userId: number,
  ): Promise<ListQuestionsResponse> {
    try {
      const questions = await this.queueService.getQuestions(queueId);
      return await this.queueService.personalizeQuestions(
        queueId,
        questions,
        userId,
        role,
      );
    } catch (err) {
      console.error(err);
      throw new HttpException(
        ERROR_MESSAGES.queueController.getQuestions,
        HttpStatus.NOT_FOUND,
      );
    }
  }

  @Patch(':queueId')
  @Roles(Role.TA, Role.PROFESSOR)
  async updateQueue(
    @Param('queueId') queueId: number,
    @Body() body: UpdateQueueParams,
  ): Promise<QueueModel> {
    const queue = await this.queueService.getQueue(queueId);
    if (queue === undefined) {
      throw new NotFoundException();
    }
    queue.notes = body.notes;
    queue.allowQuestions = body.allowQuestions;
    try {
      await queue.save();
    } catch (err) {
      console.error(err);
      throw new HttpException(
        ERROR_MESSAGES.queueController.saveQueue,
        HttpStatus.INTERNAL_SERVER_ERROR,
      );
    }
    return queue;
  }

  @Post(':queueId/clean')
  @Roles(Role.TA, Role.PROFESSOR)
  async cleanQueue(@Param('queueId') queueId: number): Promise<void> {
    // Clean up queue if necessary
    try {
      setTimeout(async () => {
        await this.queueCleanService.cleanQueue(queueId, true);
        await this.queueSSEService.updateQueue(queueId);
      });
    } catch (err) {
      console.error(err);
      throw new HttpException(
        ERROR_MESSAGES.queueController.cleanQueue,
        HttpStatus.INTERNAL_SERVER_ERROR,
      );
    }
  }

  // Endpoint to send frontend receive server-sent events when queue changes
  @Get(':queueId/sse')
  sendEvent(
    @Param('queueId') queueId: number,
    @QueueRole() role: Role,
    @UserId() userId: number,
    @Res() res: Response,
  ): void {
    res.set({
      'Content-Type': 'text/event-stream',
      'Cache-Control': 'no-cache',
      'X-Accel-Buffering': 'no',
      Connection: 'keep-alive',
    });

    try {
      this.queueSSEService.subscribeClient(queueId, res, { role, userId });
    } catch (err) {
      console.error(err);
    }
  }

  @Delete(':queueId')
  @Roles(Role.TA, Role.PROFESSOR)
  async disableQueue(
    @Param('queueId') queueId: number,
    @QueueRole() role: Role,
  ): Promise<void> {
    // disable a queue
    const queue = await this.queueService.getQueue(queueId);
    if (queue === undefined) {
      throw new NotFoundException();
    }

    if (queue.isProfessorQueue) {
      if (role == Role.TA) {
        throw new HttpException(
          ERROR_MESSAGES.queueController.cannotCloseQueue,
          HttpStatus.UNAUTHORIZED,
        );
      }
    }

    queue.isDisabled = true;

    try {
      // try to save queue
      await queue.save();
    } catch (err) {
      console.error(err);
      throw new HttpException(
        ERROR_MESSAGES.queueController.saveQueue,
        HttpStatus.INTERNAL_SERVER_ERROR,
      );
    }
  }
}
