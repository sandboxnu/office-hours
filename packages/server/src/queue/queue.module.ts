import { Module } from '@nestjs/common';
import { QueueController } from './queue.controller';
import { QueueCleanService } from './queue-clean/queue-clean.service';
import { SSEModule } from 'sse/sse.module';
import { QueueService } from './queue.service';
import { QueueSSEService } from './queue-sse.service';
import { QueueSubscriber } from './queue.subscriber';

@Module({
  controllers: [QueueController],
  providers: [
    QueueCleanService,
    QueueService,
    QueueSSEService,
    QueueSubscriber,
  ],
  exports: [QueueCleanService, QueueSSEService],
  imports: [SSEModule],
})
export class QueueModule {}
