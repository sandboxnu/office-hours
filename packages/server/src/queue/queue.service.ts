import {
  ListQuestionsResponse,
  OpenQuestionStatus,
  Question,
  QuestionGroup,
  Role,
  StatusInPriorityQueue,
  StatusInQueue,
  StatusSentToCreator,
} from '@koh/common';
import { Injectable, NotFoundException } from '@nestjs/common';
import { classToClass } from 'class-transformer';
import { pick } from 'lodash';
import { QuestionModel } from 'question/question.entity';
import { Connection, In } from 'typeorm';
import { QueueModel } from './queue.entity';
import { AlertsService } from '../alerts/alerts.service';

/**
 * Get data in service of the queue controller and SSE
 * WHY? To ensure data returned by endpoints is *exactly* equal to data sent by SSE
 */
@Injectable()
export class QueueService {
  constructor(
    private connection: Connection,
    private alertsService: AlertsService,
  ) {}

  async getQueue(queueId: number): Promise<QueueModel> {
    const queue = await QueueModel.findOne(queueId, {
      relations: ['staffList'],
    });
    await queue.checkIsOpen();
    await queue.addQueueSize();

    return queue;
  }

  async getQuestions(queueId: number): Promise<ListQuestionsResponse> {
    // todo: Make a student and a TA version of this function, and switch which one to use in the controller
    // for now, just return the student response
    const queueSize = await QueueModel.count({
      where: { id: queueId },
    });
    // Check that the queue exists
    if (queueSize === 0) {
      throw new NotFoundException();
    }

    const questionsFromDb = await QuestionModel.inQueueWithStatus(queueId, [
      ...StatusInPriorityQueue,
      ...StatusInQueue,
      OpenQuestionStatus.Helping,
    ])
      .leftJoinAndSelect('question.creator', 'creator')
      .leftJoinAndSelect('question.taHelped', 'taHelped')
      .getMany();

    const unresolvedRephraseQuestionAlerts =
      await this.alertsService.getUnresolvedRephraseQuestionAlert(queueId);

    const groupMap: Record<number, QuestionGroup> = {};

    questionsFromDb.forEach((question) => {
      if (question.groupId) {
        if (!groupMap[question.groupId]) {
          groupMap[question.groupId] = {
            id: question.groupId,
            creator: question.taHelped,
            questions: [question],
          };
        } else {
          groupMap[question.groupId].questions.push(question);
        }
      }
    });

    const questions = new ListQuestionsResponse();

    questions.queue = questionsFromDb.filter((question) =>
      StatusInQueue.includes(question.status as OpenQuestionStatus),
    );

    questions.questionsGettingHelp = questionsFromDb.filter(
      (question) =>
        question.status === OpenQuestionStatus.Helping && !question.groupId,
    );

    questions.priorityQueue = questionsFromDb.filter((question) =>
      StatusInPriorityQueue.includes(question.status as OpenQuestionStatus),
    );

    questions.groups = Object.values(groupMap);

    questions.unresolvedAlerts = unresolvedRephraseQuestionAlerts.map(
      (alert) => alert.payload,
    );

    return questions;
  }

  /** Hide sensitive data to other students */
  async personalizeQuestions(
    queueId: number,
    questions: ListQuestionsResponse,
    userId: number,
    role: Role,
  ): Promise<ListQuestionsResponse> {
    if (role === Role.STUDENT) {
      const newLQR = new ListQuestionsResponse();
      Object.assign(newLQR, questions);

      newLQR.queue = questions.queue.map((question) => {
        const creator =
          question.creator.id === userId
            ? question.creator
            : pick(question.creator, ['id']);
        // classToClass transformer will apply the @Excludes
        return classToClass<Question>(
          QuestionModel.create({ ...question, creator }),
        );
      });

      newLQR.questionsGettingHelp = questions.questionsGettingHelp.map(
        (question) => {
          const creator =
            question.creator.id === userId
              ? question.creator
              : pick(question.creator, ['id']);
          // classToClass transformer will apply the @Excludes
          return classToClass<Question>(
            QuestionModel.create({ ...question, creator }),
          );
        },
      );

      newLQR.yourQuestion = await QuestionModel.findOne({
        relations: ['creator', 'taHelped'],
        where: {
          creatorId: userId,
          queueId: queueId,
          status: In(StatusSentToCreator),
        },
      });
      newLQR.priorityQueue = [];

      return newLQR;
    }
    return questions;
  }
}
