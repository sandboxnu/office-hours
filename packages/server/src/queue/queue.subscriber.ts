import { QueueSSEService } from 'queue/queue-sse.service';
import {
  Connection,
  EntitySubscriberInterface,
  EventSubscriber,
  UpdateEvent,
} from 'typeorm';
import { QueueModel } from './queue.entity';

@EventSubscriber()
export class QueueSubscriber implements EntitySubscriberInterface<QueueModel> {
  private queueSSEService: QueueSSEService;
  constructor(connection: Connection, queueSSEService: QueueSSEService) {
    this.queueSSEService = queueSSEService;
    connection.subscribers.push(this);
  }

  // eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
  listenTo() {
    return QueueModel;
  }

  async afterUpdate(event: UpdateEvent<QueueModel>): Promise<void> {
    // Send all listening clients an update
    await this.queueSSEService.updateQueue(event.entity.id);
  }
}
