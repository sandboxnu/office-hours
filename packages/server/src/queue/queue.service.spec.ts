import { ListQuestionsResponse, QuestionStatusKeys, Role } from '@koh/common';
import { Test, TestingModule } from '@nestjs/testing';
import { mapValues, zip } from 'lodash';
import { QuestionModel } from 'question/question.entity';
import { Connection } from 'typeorm';
import {
  QuestionFactory,
  QuestionGroupFactory,
  QueueFactory,
  UserFactory,
} from '../../test/util/factories';
import { TestConfigModule, TestTypeOrmModule } from '../../test/util/testUtils';
import { QueueModel } from './queue.entity';
import { QueueService } from './queue.service';
import { AlertsService } from '../alerts/alerts.service';

describe('QueueService', () => {
  let service: QueueService;

  let conn: Connection;

  beforeAll(async () => {
    const module: TestingModule = await Test.createTestingModule({
      imports: [TestTypeOrmModule, TestConfigModule],
      providers: [QueueService, AlertsService],
    }).compile();

    service = module.get<QueueService>(QueueService);
    conn = module.get<Connection>(Connection);
  });

  afterAll(async () => {
    await conn.close();
  });

  beforeEach(async () => {
    await conn.synchronize(true);
  });

  // create 1 question for each status that exists, and put them all in a queue
  async function createQuestionsEveryStatus(
    queue: QueueModel,
  ): Promise<QuestionModel[]> {
    const allStatuses = Object.values(QuestionStatusKeys);
    const questions = await QuestionFactory.createList(allStatuses.length, {
      queue,
    });
    for (const [status, question] of zip(allStatuses, questions)) {
      question.status = status;
    }
    await QuestionModel.save(questions);
    return questions;
  }

  describe('getQuestions', () => {
    it('only returns questions in the given queue', async () => {
      const queue = await QueueFactory.create();
      await QuestionFactory.create();
      await QuestionFactory.create({ queue });
      expect((await service.getQuestions(queue.id)).queue.length).toEqual(1);
    });

    it('filters questions by status appropriately', async () => {
      const queue = await QueueFactory.create();
      await createQuestionsEveryStatus(queue);

      const questionResponse = await service.getQuestions(queue.id);
      const statuses = mapValues(
        questionResponse,
        (questions: QuestionModel[]) => questions.map((q) => q.status),
      );
      expect(statuses).toEqual({
        priorityQueue: ['PriorityQueued'],
        questionsGettingHelp: ['Helping'],
        queue: ['Queued', 'Drafting'],
        groups: [],
        unresolvedAlerts: [],
      });
    });

    it('sorts queue questions by createdat', async () => {
      const queue = await QueueFactory.create();
      const questionIds = [];
      for (let i = 0; i < 3; i++) {
        const question = await QuestionFactory.create({
          queue,
          createdAt: new Date(Date.now() + i * 1000),
        });
        questionIds.push(question.id);
      }

      expect(
        (await service.getQuestions(queue.id)).queue.map((q) => q.id),
      ).toEqual(questionIds);
    });

    it('sorts priority queue questions by createdat', async () => {
      const queue = await QueueFactory.create();
      const questionIds = [];
      for (let i = 0; i < 3; i++) {
        const question = await QuestionFactory.create({
          queue,
          status: 'PriorityQueued',
          createdAt: new Date(Date.now() + i * 1000),
        });
        questionIds.push(question.id);
      }

      expect(
        (await service.getQuestions(queue.id)).priorityQueue.map((q) => q.id),
      ).toEqual(questionIds);
    });

    it('fetches questions in all groups', async () => {
      const queue = await QueueFactory.create();
      await createQuestionsEveryStatus(queue);
      const group1 = await QuestionGroupFactory.create({ queue });
      const g1q1 = await QuestionFactory.create({
        queue,
        groupable: true,
        group: group1,
      });
      const g1q2 = await QuestionFactory.create({
        queue,
        groupable: true,
        group: group1,
      });
      const group2 = await QuestionGroupFactory.create({ queue });
      const g2q1 = await QuestionFactory.create({
        queue,
        groupable: true,
        group: group2,
      });

      const recievedGroups = (await service.getQuestions(queue.id)).groups;

      expect(recievedGroups.length).toEqual(2);
      recievedGroups.forEach((group) => {
        if (group.id === group1.id) {
          expect(group.questions.length).toEqual(2);
          expect(group.questions.some((q) => q.id === g1q1.id)).toBeTruthy();
          expect(group.questions.some((q) => q.id === g1q2.id)).toBeTruthy();
        } else {
          expect(group.questions.length).toEqual(1);
          expect(group.questions.some((q) => q.id === g2q1.id)).toBeTruthy();
        }
      });
    });
  });

  describe('personalizeQuestions', () => {
    let queue;
    beforeEach(async () => {
      queue = await QueueFactory.create();
    });
    const personalize = (
      lqr: ListQuestionsResponse,
      userId: number,
      role: Role,
    ) => service.personalizeQuestions(queue.id, lqr, userId, role);

    it('does nothing if not a student', async () => {
      const user = await UserFactory.create();
      await QuestionFactory.create({
        queue,
        createdAt: new Date('2020-11-02T12:00:00.000Z'),
      });
      const lqr = await service.getQuestions(queue.id);
      expect(
        await service.personalizeQuestions(queue.id, lqr, user.id, Role.TA),
      ).toEqual(lqr);
    });

    it('adds yourQuestion for students with question in the queue', async () => {
      const user = await UserFactory.create();
      // Create a question but not in this queue
      await QuestionFactory.create({ creator: user });

      const blank: ListQuestionsResponse = {
        queue: [],
        priorityQueue: [],
        questionsGettingHelp: [],
        groups: [],
      };
      let lqr = await personalize(blank, user.id, Role.STUDENT);
      expect(lqr.yourQuestion).toEqual(undefined);

      // Create a question in this queue
      const question = await QuestionFactory.create({
        creator: user,
        queue,
      });
      lqr = await personalize(blank, user.id, Role.STUDENT);
      expect(lqr.yourQuestion.id).toEqual(question.id);
    });

    it('hides details of other students', async () => {
      const ours = await QuestionFactory.create({
        queue,
        createdAt: new Date('2020-11-02T12:00:00.000Z'),
        text: 'help us',
      });
      await QuestionFactory.create({
        queue,
        createdAt: new Date('2020-11-02T12:00:00.000Z'),
        text: 'help someone else',
      });

      const lqr = await personalize(
        await service.getQuestions(queue.id),
        ours.creatorId,
        Role.STUDENT,
      );
      expect(lqr).toMatchSnapshot();
    });
  });
});
