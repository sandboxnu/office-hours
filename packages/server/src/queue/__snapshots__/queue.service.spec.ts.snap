// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QueueService personalizeQuestions hides details of other students 1`] = `
ListQuestionsResponse {
  "groups": Array [],
  "priorityQueue": Array [],
  "questionsGettingHelp": Array [],
  "queue": Array [
    QuestionModel {
      "closedAt": null,
      "createdAt": 2020-11-02T12:00:00.000Z,
      "creator": UserModel {
        "defaultMessage": null,
        "email": "user@neu.edu",
        "firstName": "User",
        "id": 1,
        "includeDefaultMessage": true,
        "lastName": "Person",
        "photoURL": null,
      },
      "groupable": true,
      "helpedAt": null,
      "id": 1,
      "location": null,
      "questionType": "Other",
      "status": "Queued",
      "taHelped": null,
      "text": "help us",
    },
    QuestionModel {
      "closedAt": null,
      "createdAt": 2020-11-02T12:00:00.000Z,
      "creator": UserModel {
        "id": 2,
      },
      "groupable": true,
      "helpedAt": null,
      "id": 2,
      "location": null,
      "questionType": "Other",
      "status": "Queued",
      "taHelped": null,
      "text": "help someone else",
    },
  ],
  "unresolvedAlerts": Array [],
  "yourQuestion": QuestionModel {
    "closedAt": null,
    "createdAt": 2020-11-02T12:00:00.000Z,
    "creator": UserModel {
      "defaultMessage": null,
      "desktopNotifsEnabled": false,
      "email": "user@neu.edu",
      "firstName": "User",
      "hideInsights": Array [],
      "id": 1,
      "includeDefaultMessage": true,
      "insights": Array [
        "TotalStudents",
        "TotalQuestionsAsked",
        "MedianWaitTime",
        "QuestionTypeBreakdown",
        "MostActiveStudents",
        "QuestionToStudentRatio",
        "MedianHelpingTime",
      ],
      "lastName": "Person",
      "name": "User Person",
      "phoneNotifsEnabled": false,
      "photoURL": null,
    },
    "creatorId": 1,
    "firstHelpedAt": null,
    "groupId": null,
    "groupable": true,
    "helpedAt": null,
    "id": 1,
    "location": null,
    "questionType": "Other",
    "queueId": 1,
    "status": "Queued",
    "taHelped": null,
    "taHelpedId": null,
    "text": "help us",
  },
}
`;
